Add UF2 output file generation

petejohanson opened the pull request:
Various devices are now shipping w/ UF2 compatible bootloaders. We've been hacking this into Zephyr ourselves w/ per-board post build commands for ZMK, but this adds proper upstream support for building UF2 files as part of the output, and adds support for the `seeeduino_xiao` board which has the BOSSA UF2 bootloader.

Tested locally building the `samples/basic/blinky` sample, double tapping reset, then `cp build/xiao/zephyr/zephyr.uf2 /run/media/peter/Arduino` and it is happily blinking away.

Adds new `BUILD_OUTPUT_UF2_FAMILY_ID` Kconfig setting for the family ID needed: https://github.com/Microsoft/uf2#family-id

The only thing I have an open question about would be the `west.yml` changes. I assume before merging the standard is to fork the upstream project into `zephyrproject-rtos`?

TIA!

petejohanson committed:
cmake: Add UF2 as an additional build output option.

Add ability to build a UF2 (https://github.com/Microsoft/uf2)
image as an additional output type. This leverages the code
partition offset for the UF2 base address, and a configurable
UF2 family ID.

Includes an unmodified (except for headers for licensing, pylit
disabling) version of the uf2conv.py script copied the UF2
format specification repository, used to convert the bin to UF2.

Origin: UF2 file format specification reference utilies
License: MIT
URL: https://github.com/microsoft/uf2/blob/master/utils/uf2conv.py
commit: 587abb8b909266e9b468d6284f2fbd425235d1b5
Signed-off-by: Pete Johanson <peter@peterjohanson.com>

tejlmand commented on CMakeLists.txt:
```diff
@@ -1159,6 +1159,21 @@ if(CONFIG_BUILD_OUTPUT_BIN)
   endif()
 endif()
 
+if(CONFIG_BUILD_OUTPUT_BIN AND CONFIG_BUILD_OUTPUT_UF2)
```
I think UF2 deserves to be a Zephyr module of its own, and therefore this code should go into the CMakeLists.txt code for that module.

This will also ensure the options are only available when the module is available.

petejohanson replied:
Making sure I understand this correctly... When #30904 is merged, the only thing in the actual external module needed will be the `zephyr/module.yml`, and then the `CMakeLists.txt` and `Kconfig.uf2` files would be in the main Zephyr repo, and only included if that uf2 module was also fetched by west and was present?

tejlmand replied:
that is correctly understood.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -318,6 +318,25 @@ config BUILD_OUTPUT_BIN
 	  Build a "raw" binary zephyr/zephyr.bin in the build directory.
 	  The name of this file can be customized with CONFIG_KERNEL_BIN_NAME.
 
+config BUILD_OUTPUT_UF2
```
Same as with CMake code.
Should be part of the module, and thus be in `modules/Kconfig.uf2`.

tejlmand commented on west.yml:
```diff
@@ -144,6 +146,10 @@ manifest:
       revision: 1.7.0-rc1
+    - name: uf2
+      repo-path: uf2
+      remote: microsoft
```
would need a fork into Zephyr organization, as described:
https://docs.zephyrproject.org/latest/guides/modules.html#module-repositories

- All modules included in the default manifest shall be hosted in repositories under the zephyrproject-rtos GitHub organization.
- The module repository codebase shall include a module.yml file in a zephyr/ folder at the root of the repository.
- ...

tejlmand replied:
as well as this:
```suggestion
```

tejlmand reviewed (changes requested):
Thanks for this PR.

> The only thing I have an open question about would be the west.yml changes. I assume before merging the standard is to fork the upstream project into zephyrproject-rtos?

That is correct.
You can find some more information here:
https://docs.zephyrproject.org/latest/guides/modules.html#module-repositories

So we need a fork @nashif and @carlescufi can you take care of that ?

And in that fork we would need a `zephyr/module.yml` and a CMakeLists.txt file.

If you wait for: #30904, then you can place all Zephyr glue code in the main Zephyr repo, and only the `zephyr/module.yml` file will be needed in the fork.

pabigot commented:
AFAICT this would enable Zephyr on a large number of [Adafruit Nordic boards](https://github.com/adafruit/Adafruit_nRF52_Bootloader) @carlescufi.

If true, I'd very much like to see this merged.  Let me know if there's something I can do to help.

petejohanson commented:
> AFAICT this would enable Zephyr on a large number of [Adafruit Nordic boards](https://github.com/adafruit/Adafruit_nRF52_Bootloader) @carlescufi.
> 
> If true, I'd very much like to see this merged. Let me know if there's something I can do to help.

That is correct. This is based on our initial hacky post build command work we've done in https://github.com/zmkfirmware/zmk for now to support boards using that exact bootloader. Trying to "do this right" by getting this merged into Zephyr proper.

pabigot commented:
Will review after @tejlmand's blocking change requests are addressed, as testing before that won't be much use.

petejohanson commented:
My plan is to rework this once #30904 is in.

carlescufi commented:
> My plan is to rework this once #30904 is in.

It's in now @petejohanson

tejlmand commented:
> > My plan is to rework this once #30904 is in.
> 
> It's in now @petejohanson

@carlescufi I think we also need a Zephyr fork, as mentioned: https://github.com/zephyrproject-rtos/zephyr/pull/31066#pullrequestreview-566146871

petejohanson commented:
@tejlmand @pabigot @carlescufi @pabigot 

Ok, I've just reworked this based on the new modules work.

Since we don't yet have the fork of the uf2 repo, the `west.yml` changes are still as is.

This assumes the following:

1. Checkout this PR.
2. Run `west update`
3. `mkdir tools/uf2/zephyr`
4. Add `tools/uf2/zephyr/module.yml` file with `cmake-ext` and `kconfig-ext` set to `True`
5. Build w/ the `seeeduino_xiao`, e.g. `west build -d build/xiao -b seeeduino_xiao samples/basic/blinky`

I've only updated the XIAO for this, since it is one known good board in upstream Zephyr that comes OOTB w/ a UF2 enabled bootloader.

If everyone is happy, and can get the uf2 repo forked into Zephyr's org, I can update the `west.yml` appropriately.

Thanks again!

petejohanson commented:
Any thoughts on this?

pabigot commented:
> Any thoughts on this?

A lot of us are pretty focused on the stabilization phase for 2.5 right now, so we may not get to this until 2.5.0 closer to being ready for the release in the second week of February.  (At least, I probably won't look at this until then.)

carlescufi commented:
> Any thoughts on this?

Could you please state which repository exactly you need forked, what it provides to Zephyr and which changes do you expect to have in it?

petejohanson commented:
> > Any thoughts on this?
> 
> Could you please state which repository exactly you need forked, what it provides to Zephyr and which changes do you expect to have in it?

Repo: https://github.com/microsoft/uf2
Why needed: It contains a python script, https://github.com/microsoft/uf2/blob/master/utils/uf2conv.py that can convert a bin/hex to a UF2 file. This is as authoritative a conversion tool that I've found, and being python, is easy for us to use.

The only modification needed is the creation of one file `zephyr/module.yml` which should contain;

```
build:
  cmake-ext: True
  kconfig-ext: True
```

This is based on the work in #30904.

With this in place, the extra CMakeLists.txt added in this PR will work along w/ the few extra Kconfig settings, to build UF2 files as an extra post build step.

tejlmand commented on modules/uf2/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+if(CONFIG_BUILD_OUTPUT_BIN AND CONFIG_BUILD_OUTPUT_UF2)
+  set_property(GLOBAL APPEND PROPERTY
+    extra_post_build_commands
+    COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/../tools/uf2/utils/uf2conv.py
```
When we have this as a Zephyr module, you should use:
```suggestion
    COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_UF2_MODULE_DIR}/utils/uf2conv.py
```

tejlmand commented on west.yml:
```diff
@@ -21,6 +21,8 @@ manifest:
   remotes:
     - name: upstream
       url-base: https://github.com/zephyrproject-rtos
+    - name: microsoft
+      url-base: https://github.com/microsoft
```
when the fork is in place, this can be removed.
```suggestion
```

tejlmand reviewed (changes requested):
Thanks for this.

Looks good, only a couple of small cleanups needed when the fork is in place.

tejlmand commented:
@carlescufi  please make the `zephyr/module.yml` file as:
```
name: uf2

build:
  cmake-ext: True
  kconfig-ext: True
```

carlescufi commented:
> Repo: https://github.com/microsoft/uf2
> Why needed: It contains a python script, https://github.com/microsoft/uf2/blob/master/utils/uf2conv.py that can convert a bin/hex to a UF2 file. This is as authoritative a conversion tool that I've found, and being python, is easy for us to use.

Thanks for explaining. But does it make sense to fork a whole repo just for that single scripts in a `utils/` folder?

petejohanson commented:
I don't particularly care. The only advantage I would see would be easier integration of future changes upstream. If you take a look, that script had updates 15 days ago to add the RP2040 family, so changes to support more families would require occasional updates to the version we integration into Zephyr.

Whether those updates take the form of a repo fork, or a copy of a single file into Zephyr, I'll defer to others.

carlescufi commented:
> I don't particularly care. The only advantage I would see would be easier integration of future changes upstream. If you take a look, that script had updates 15 days ago to add the RP2040 family, so changes to support more families would require occasional updates to the version we integration into Zephyr.
> 
> Whether those updates take the form of a repo fork, or a copy of a single file into Zephyr, I'll defer to others.

OK. I will mark this particular PR with the TSC label, since we are coincidentally discussing this very matter there.

galak commented:
Any possibility of asking microsoft to stick this in pip?

petejohanson commented:
> Any possibility of asking microsoft to stick this in pip?
> 
> [microsoft/uf2#42](https://github.com/microsoft/uf2/issues/42)

Yeah, this would be a much nicer way of handling this. Curious what they say. Given the 2.5 timeframe... Is there some temporary measure possible pending pip publish? (those 'p's were fun to write!)

galak commented:
> > [](https://github.com/microsoft/uf2/issues/42)
> 
> Yeah, this would be a much nicer way of handling this. Curious what they say. Given the 2.5 timeframe... Is there some temporary measure possible pending pip publish? (those 'p's were fun to write!)

This is already too late for 2.5 at this point.  We're in bug fix only phase.

mbolivar-nordic commented on modules/uf2/Kconfig:
```diff
@@ -0,0 +1,21 @@
+
+config BUILD_OUTPUT_UF2
+	bool "Build a binary in UF2 format"
+	default n
```
```suggestion
```

unnecessary `default n` nit

tejlmand commented:
> Thanks for explaining. But does it make sense to fork a whole repo just for that single script in a `utils/` folder? It's also a pretty simple script, why not just take it and place it in our `scripts/` folder in the main tree?

Good point, I didn't notice that it's only a single file we need for this purpose.
In that case, if this gets solved https://github.com/microsoft/uf2/issues/42 that would be a much nicer approach, and until then a local copy of the script under `scripts` might be the cleanest solution, as there will be no need for creation of an extra project in peoples west workspaces, just to end with a dangling repo when later switching to PyPI and that repo is removed from `west.yml`.

petejohanson commented:
> > Thanks for explaining. But does it make sense to fork a whole repo just for that single script in a `utils/` folder? It's also a pretty simple script, why not just take it and place it in our `scripts/` folder in the main tree?
> 
> Good point, I didn't notice that it's only a single file we need for this purpose.
> In that case, if this gets solved [microsoft/uf2#42](https://github.com/microsoft/uf2/issues/42) that would be a much nicer approach, and until then a local copy of the script under `scripts` might be the cleanest solution, as there will be no need for creation of an extra project in peoples west workspaces, just to end with a dangling repo when later switching to PyPI and that repo is removed from `west.yml`.

I'm happy to do this. If we go this route, can I just integrate the build changes into the core CMake stuff as it was before, and skip the module approach?

tejlmand commented:
> I'm happy to do this. If we go this route, can I just integrate the build changes into the core CMake stuff as it was before, and skip the module approach?

Correct, and thank you for your patience.

carlescufi commented:
> > Any possibility of asking microsoft to stick this in pip?
> > [microsoft/uf2#42](https://github.com/microsoft/uf2/issues/42)
> 
> Yeah, this would be a much nicer way of handling this. Curious what they say. Given the 2.5 timeframe... Is there some temporary measure possible pending pip publish? (those 'p's were fun to write!)

@petejohanson In the meantime, could you please remove the changes to `west.yml` and import the script directly in `scripts/`?

carlescufi commented:
@petejohanson we agreed in the TSC to move forward with this with the `uf2conv.py` imported manually into zephyr's `scripts/` folder.

petejohanson commented:
@tejlmand @carlescufi Done and done, Only modifications I made to the script was to add a license identifier comment, and a comment indicating the exact sha1 and location where the script was copied from.

Thanks again.

galak commented on CMakeLists.txt:
```diff
@@ -1159,6 +1159,21 @@ if(CONFIG_BUILD_OUTPUT_BIN)
   endif()
 endif()
 
+if(CONFIG_BUILD_OUTPUT_BIN AND CONFIG_BUILD_OUTPUT_UF2)
```
Why do we need both conditions in the if.  How can UF2 be set w/o BIN?

petejohanson replied:
It can't really, since the Kconfig depends on it. I can change the conditional.

petejohanson replied:
The reason I'd added was to be explicit, since we invoke and pass in the bin output to the script.

petejohanson commented:
So... is there a way to make the compliance the license checks happy with this? 

I've just read through https://docs.zephyrproject.org/latest/contribute/index.html#identifying-contribution-origin and updated the commit that brings in `uf2conv.py` to hopefully have all the correct metadata in the commit body.

tejlmand commented:
> So... is there a way to make the compliance the license checks happy with this?

Add an entry to [CODEOWNERS](https://github.com/zephyrproject-rtos/zephyr/blob/master/CODEOWNERS) for this file:
> ERROR   : Test Codeowners failed: New files added that are not covered in CODEOWNERS:
>
> scripts/uf2conv.py

Maybe disable the pylint warnings using `# pylint: disable=<warning>` , as this script originates from another place.
> ERROR   : Test pylint failed: ************* Module uf2conv
> Running pylint           tests in /home/runner/work/zephyr/zephyr ...
> scripts/uf2conv.py:203:0: W1401: Anomalous backslash in string: '\s'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
> scripts/uf2conv.py:84:11: C0121: Comparison to None should be 'expr is None' (singleton-comparison)
2 checks failed
> scripts/uf2conv.py:104:4: C0200: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
>
> scripts/uf2conv.py:177:15: C0121: Comparison to None should be 'expr is None' (singleton-comparison)
> scripts/uf2conv.py:221:8: W0702: No exception type(s) specified (bare-except)
> scripts/uf2conv.py:306:15: C0121: Comparison to None should be 'expr is None' (singleton-comparison)

tejlmand commented on Kconfig.zephyr:
```diff
@@ -344,6 +344,25 @@ config BUILD_OUTPUT_S19
 
+config BUILD_OUTPUT_UF2
+	bool "Build a binary in UF2 format"
+	default n
```
Default is `n` so just remove.
```suggestion
```

petejohanson replied:
Removed. Thanks, I missed that earlier review comment.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -344,6 +344,25 @@ config BUILD_OUTPUT_S19
+	string "UF2 Device Family"
+	help
+	  UF2 bootloaders only accept UF2 files with a matching family ID.
+	  This can be either a hex, e.g. 0x68ed2b88, or well-known family
+	  name string.
```
Based on this link: https://github.com/Microsoft/uf2#picking-numbers-at-random
I think it should be highlighted that it should not just be some hex numbers, but a good random number, and if a family name is used, then the known family name will be converted to the corresponding hex value.

Should we lock this field to a specific value if a SoC with a know family ID is used ?
Such as the nrf52, samd51, stm32, etc.

petejohanson replied:
I'm not sure about locking this field, but maybe *defaulting* the family ID value on those SoCs? If there's interest, I can update this to do that.

petejohanson replied:
@tejlmand Ok, I've just rebased and tweaked this a bit, and have added defaulting of the family ID config for the various SoC defconfig files, so for a given board, like the Seeeduino Xiao now in this PR, just enabling the UF2 build output is enough.

galak commented:
> So... is there a way to make the compliance the license checks happy with this?

You can just have someone with merge rights merge and not worry about the compliance issues with the python script.

petejohanson commented:
> > So... is there a way to make the compliance the license checks happy with this?
> 
> Add an entry to [CODEOWNERS](https://github.com/zephyrproject-rtos/zephyr/blob/master/CODEOWNERS) for this file:
> 
> > ERROR   : Test Codeowners failed: New files added that are not covered in CODEOWNERS:
> > scripts/uf2conv.py
> 
> Maybe disable the pylint warnings using `# pylint: disable=<warning>` , as this script originates from another place.
> 
> > ERROR   : Test pylint failed: ************* Module uf2conv
> > Running pylint           tests in /home/runner/work/zephyr/zephyr ...
> > scripts/uf2conv.py:203:0: W1401: Anomalous backslash in string: '\s'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
> > scripts/uf2conv.py:84:11: C0121: Comparison to None should be 'expr is None' (singleton-comparison)
> > 2 checks failed
> > scripts/uf2conv.py:104:4: C0200: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
> > scripts/uf2conv.py:177:15: C0121: Comparison to None should be 'expr is None' (singleton-comparison)
> > scripts/uf2conv.py:221:8: W0702: No exception type(s) specified (bare-except)
> > scripts/uf2conv.py:306:15: C0121: Comparison to None should be 'expr is None' (singleton-comparison)

Done, and done. Thanks.

petejohanson committed:
boards: seeeduino_xiao: Output UF2 by default.

Default BOSSA bootloader supports UF2 OOTB, so enabling
UF2 output by default.

Signed-off-by: Pete Johanson <peter@peterjohanson.com>

mbolivar-nordic commented:
@petejohanson nice!

Can I suggest you consider adding a [mass storage runner](https://github.com/zephyrproject-rtos/zephyr/issues/12985) which takes advantage of this so `west flash` does the right thing here? I haven't had time to do this and I think it'd be pretty cool.

You would need to add a `uf2_file` field in the RunnerConfig tuple in zephyr/scripts/west_commands/runners/core.py, then add a runner which copies that file to the "right" mass storage device.

I don't know how these devices appear on the platforms you're interested in. I would hope there'd be a cross-platform way to discover the potential ones at runtime and either flash to the only one that's present, or prompt the user to pick one, similarly to now the nrfjprog runner runs `nrfjprog --ids` to figure out which J-Link devices are candidates for flashing an nRF board.

Just spitballing potential future improvements. I am definitely not trying to block this PR, which I know you've been patiently waiting for merge on for a while.

petejohanson commented:
> @petejohanson nice!
> 
> Can I suggest you consider adding a [mass storage runner](https://github.com/zephyrproject-rtos/zephyr/issues/12985) which takes advantage of this so `west flash` does the right thing here? I haven't had time to do this and I think it'd be pretty cool.

Yes, this has definitely been on my radar as a next step.

> 
> You would need to add a `uf2_file` field in the RunnerConfig tuple in zephyr/scripts/west_commands/runners/core.py, then add a runner which copies that file to the "right" mass storage device.
> 
> I don't know how these devices appear on the platforms you're interested in. I would hope there'd be a cross-platform way to discover the potential ones at runtime and either flash to the only one that's present, or prompt the user to pick one, similarly to now the nrfjprog runner runs `nrfjprog --ids` to figure out which J-Link devices are candidates for flashing an nRF board.

Yeah, I've been pondering this aspect too. On my Fedora laptop, they end up auto mounted at `/run/media/peter/FOO`, Windows some device letter, macOS not sure. Hoping maybe there is an existing python library I could leverage.


> Just spitballing potential future improvements. I am definitely not trying to block this PR, which I know you've been patiently waiting for merge on for a while.

No worries!

tejlmand commented on Kconfig.zephyr:
```diff
@@ -344,6 +344,24 @@ config BUILD_OUTPUT_S19
+	help
+	  UF2 bootloaders only accept UF2 files with a matching family ID.
+	  This can be either a hex, e.g. 0x68ed2b88, or well-known family
+	  name string.
```
```suggestion
	  name string. If the SoC in use is known by UF2, the Family ID will be pre-filled with the known value.
```

tejlmand commented on Kconfig.zephyr:
```diff
@@ -344,6 +344,24 @@ config BUILD_OUTPUT_S19
+if BUILD_OUTPUT_UF2
+
+config BUILD_OUTPUT_UF2_FAMILY_ID
+	string "UF2 Device Family"
```
I suggest to create a default list in `modules/Kconfig.uf2` as shown below and then add a line in https://github.com/zephyrproject-rtos/zephyr/blob/master/modules/Kconfig
```
source "modules/Kconfig.uf2"
```

```
config BUILD_OUTPUT_UF2_FAMILY_ID
	# Prompt less config to setup a list of default UF2 Family IDs 
	default "0x68ed2b88" if SOC_SERIES_SAMD21
	default "0x55114460" if SOC_SERIES_SAMD51
	default "0xada52840" if SOC_NRF52840_QIAA
	default "0x4fb2d5bd" if SOC_SERIES_IMX_RT
	default "0x2abc77ec" if SOC_SERIES_LPC55XXX
	default "0x647824b6" if SOC_SERIES_STM32F0X
	default "0x5ee21072" if SOC_STM32F103XE
	default "0x5d1a0a2e" if SOC_SERIES_STM32F2X
	default "0x6b846188" if SOC_SERIES_STM32F3X
	default "0x57755a57" if SOC_STM32F401XC
	default "0x57755a57" if SOC_STM32F401XE
	default "0x6d0922fa" if SOC_STM32F407XE
	default "0x8fb060fe" if SOC_STM32F407XG
	default "0x53b80f00" if SOC_SERIES_STM32F7X
	default "0x300f5633" if SOC_SERIES_STM32G0X
	default "0x4c71240a" if SOC_SERIES_STM32G4X
	default "0x6db66082" if SOC_SERIES_STM32H7X
	default "0x202e3a91" if SOC_SERIES_STM32L0X
	default "0x1e1f432d" if SOC_SERIES_STM32L1X
	default "0x00ff6919" if SOC_SERIES_STM32L4X
	default "0x04240bdf" if SOC_SERIES_STM32L5X
	default "0x70d16653" if SOC_SERIES_STM32WBX
	default "0x1c5f21b0" if SOC_ESP32
	help
	  Default UF2 Family IDs for known SoCs.
	  All values on this list must correspond to the values found in  scripts/uf2conv.py.
```

This allows us to not pollute the `Kconfig.zephyr` but still have a common default list of known boards without updating each board file.

**Note**: I propose `modules/Kconfig.uf2` although this is strictly not module code.
But I feel `subsys` or `lib` are worse alternative, and as there is no actual config settings (only default values), I feel this is best place.
Better proposals are welcome.

pabigot replied:
`soc/Kconfig{,.uf2}` if we want them all in one place.

Does:

    default "0x57755a57" if SOC_STM32F401XC || SOC_STM32F401XE

work?  And sort them, either by value or by `SOC_` selector.

tejlmand replied:
> Does:
> 
> ```
> default "0x57755a57" if SOC_STM32F401XC || SOC_STM32F401XE
> ```
> 
> work? And sort them, either by value or by `SOC_` selector.

It sure does, I didn't notice there were duplicated default values.
Sorting is a good idea, so I updated proposal above accordingly, and I think based on SOC makes most sense, as the default value is supposed to be a random number.

Will update proposal accordingly.

> `soc/Kconfig{,.uf2}` if we want them all in one place.

Thanks for the proposal. This is also a possibility.
I think i'm split between this and the `modules`.
Small recap:
- `modules/Kconfig.uf2` clearly indicates that this is unrelated to core Zephyr SoC in general, like other module code, cons is that this is not pure module code.
- `soc/Kconfig.soc` indicates that the default settings for UF2 Family ID is related to SoC names, but i'm unsure if this will be first place people will look if looking for default `BUILD_OUTPUT_UF2_FAMILY_ID` values.

tejlmand reviewed (changes requested):
Thanks for the update, and for going through the SoCs already know byt UF2.
Makes it much easier to make the minor default proposal update.

I propose to keep the knowledge of Zephyr SoC to UF2 Family ID mapping close together, similar to what is done for TF-M.

This remove the need for SoCs to know anything about UF2.

pabigot commented on soc/arm/atmel_sam0/samd21/Kconfig.defconfig.series:
```diff
@@ -30,4 +30,11 @@ config NUM_IRQS
 config SYS_CLOCK_HW_CYCLES_PER_SEC
 	default 48000000
 
+if BUILD_OUTPUT_UF2
+
+config BUILD_OUTPUT_UF2_FAMILY_ID
+	default "0x68ed2b88"
+
+endif
+
```
Retargeting the discussion from https://github.com/zephyrproject-rtos/zephyr/pull/31066#discussion_r576094245: on further review I now completely disagree with the suggestion to move these out to a single list maintained separately.

The UF2 IDs appear to come [from the uf2 script](https://github.com/microsoft/uf2/blob/6e0f282d4b1a56548e4780de95498e9de158fe37/utils/uf2conv.py#L15) and are associated with specific SoC families.

So I think these belong exactly here, in the SoC Kconfig, and that the value for `BUILD_OUTPUT_UF2_FAMILY_ID` should be the text identifier for the family, in this case `"SAMD21"`, rather than the hex constant.

Then when the script is updated to support more SoCs the corresponding SoC Kconfigs should have the property added.

Also since there's only one default here it seems cleaner to use:

```
config BUILD_OUTPUT_UF2_FAMILY_ID
	depends on BUILD_OUTPUT_UF2
	default "SAMD21"
```

rather than an enclosing conditional, especially since that is the [underlying equivalent](https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#factoring-out-common-dependencies).

tejlmand replied:
> Retargeting the discussion from [#31066 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/31066#discussion_r576094245): on further review I now completely disagree with the suggestion to move these out to a single list maintained separately.
> 
> The UF2 IDs appear to come [from the uf2 script](https://github.com/microsoft/uf2/blob/6e0f282d4b1a56548e4780de95498e9de158fe37/utils/uf2conv.py#L15) and are associated with specific SoC families.

I disagree there.
This is no different than drivers Kconfig, Bluetooth, TF-M where it is no code owners responsible for those areas who will have defaults in their Kconfig files.
This also goes for boards.

All such knowledge doesn't go into boards or soc Kconfigs, as this will become unmaintainable.
https://github.com/zephyrproject-rtos/zephyr/blob/a890790592d85d7804f5059618d94e1acaeebe5a/drivers/flash/Kconfig.stm32#L7-L32

https://github.com/zephyrproject-rtos/zephyr/blob/a890790592d85d7804f5059618d94e1acaeebe5a/modules/Kconfig.tfm#L8-L16

and many more.

With your rationale, anything that has a default that is based on a specific SoC could then be placed in the corresponding Kconfig soc file.

A SoC maintainer should not need to consider UF2, this is for whoever maintains the UF2 feature, and I think it's UF2 is to be seen as independent feature.

pabigot replied:
OK, that's somewhat compelling, but then the Kconfig defaults should go into the board Kconfig.defconfig for boards that are likely to support UF2 bootloaders, just as Nordic default-enables `BT_CTLR` for their boards.

This isn't using a module, so its configuration doesn't belong in `modules/`.  Either put it as defaults where the variable is defined, or put it in the board Kconfig.defconfig.  Hiding it elsewhere just makes it painful.

pabigot reviewed (changes requested):
I think this is a good enhancement that will make Adafruit boards a lot easier to use in Zephyr.

Please fix the commit messages:

    feat(build): Add UF2 family IDs for registered SoCs.

is not how we format these.

    soc: Add UF2 family IDs for supported SoCs

It's also unusual to make the commit messages lists of changes.  Free-form text is standard.

The `Purpose:` and `Maintained-By:` lines should be removed from the `cmake: ` commit; the purpose should be in the description, and once it's in Zephyr its Zephyr that maintains it.  The commit message could/should document that the expectation is this is unmodified from the source repository version.

petejohanson commented:
> I think this is a good enhancement that will make Adafruit boards a lot easier to use in Zephyr.

There's quite a few other boards supporting UF2 natively now, including the RP2040, which support UF2 from their default/onboard bootloader.

> 
> Please fix the commit messages:
> 
>     feat(build): Add UF2 family IDs for registered SoCs.
> 
> is not how we format these.
> 
>     soc: Add UF2 family IDs for supported SoCs
> 
> It's also unusual to make the commit messages lists of changes.  Free-form text is standard.
> 
> The `Purpose:` and `Maintained-By:` lines should be removed from the `cmake: ` commit; the purpose should be in the description, and once it's in Zephyr its Zephyr that maintains it.  The commit message could/should document that the expectation is this is unmodified from the source repository version.

Will do.

pabigot commented:
> > I think this is a good enhancement that will make Adafruit boards a lot easier to use in Zephyr.
> 
> There's quite a few other boards supporting UF2 natively now, including the RP2040, which support UF2 from their default/onboard bootloader.

OK.  BTW my block is solely on this stuff:
 
> > Please fix the commit messages:
ory version.
> 
> Will do.

For the other point of contention (defaulting `BUILD_OUTPUT_UF2_FAMILY_ID` based on SOC) , based on all discussion so far just putting the defaults as text identifiers in `Kconfig.zepyhyr` where `config BUILD_OUTPUT_UF2_FAMILY_ID` is introduced seems the most maintainable.  It's not like it's hundreds of lines.

But if it has to be squirreled away in an include file under `modules/` that's not something I'd block on.  (I would block on using the hex constants instead of the text identifiers from the uf2 script, unless we actually need the hex constants explicitly.)

petejohanson commented:
> 
> But if it has to be squirreled away in an include file under `modules/` that's not something I'd block on.  (I would block on using the hex constants instead of the text identifiers from the uf2 script, unless we actually need the hex constants explicitly.)
> 

I was going to comment on this. It seems like going with the family name alias, instead of the actual hex value ties us to an implementation detail of `uf2conv.py`, which seemed good to avoid. The actual hex value seems to be a fixed thing, is the friendly name alias?

I don't care strongly either way, just a thought.

Should a wait for concensus on where to put the defaults?

pabigot commented:
> I was going to comment on this. It seems like going with the family name alias, instead of the actual hex value ties us to an implementation detail of `uf2conv.py`, which seemed good to avoid. The actual hex value seems to be a fixed thing, is the friendly name alias?

Is there a registry of known identifiers that is more canonical than the `uf2conv` script?  In a quick search I didn't find one.

Is there an alternative to `uf2conv.py` that has any market share?

If the answers are "no" and "no", then for now I'd say that script is the reference definition, and using the names it supports should be pretty stable.

> Should a wait for concensus on where to put the defaults?

Couldn't say.  I'm not sure if @tejlmand is open to anything other than what he's proposed, but that solution is something I'm willing to accept.  So I guess just run with it and see if we can get this merged before my Clue boards arrive this week.

petejohanson commented:
> > I was going to comment on this. It seems like going with the family name alias, instead of the actual hex value ties us to an implementation detail of `uf2conv.py`, which seemed good to avoid. The actual hex value seems to be a fixed thing, is the friendly name alias?
> 
> Is there a registry of known identifiers that is more canonical than the `uf2conv` script? In a quick search I didn't find one.

Actually, yes. The toplevel README.md in the uf2 repository lists *more* chosen IDs than have aliases in the `uf2conv.py` script. In particular, the ESP32 family ID (`0x1c5f21b0`) doesn't have a corresponding alias. This may be oversight, on the part of the maintainers there, but it points to a bit of a disconnect.

> 
> Is there an alternative to `uf2conv.py` that has any market share?

No that I've seen, but I haven't gone looking.

> 
> If the answers are "no" and "no", then for now I'd say that script is the reference definition, and using the names it supports should be pretty stable.
> 
> > Should a wait for concensus on where to put the defaults?
> 
> Couldn't say. I'm not sure if @tejlmand is open to anything other than what he's proposed, but that solution is something I'm willing to accept. So I guess just run with it and see if we can get this merged before my Clue boards arrive this week.

I'm going to update to put them inline in `Kconfig.zephyr` to start, it's easy enough to then move that to a separate (single) file if requested.

petejohanson committed:
soc: Add standard UF2 family IDs for supported SoCs.

Defaults from https://github.com/Microsoft/uf2#family-list

Signed-off-by: Pete Johanson <peter@peterjohanson.com>

petejohanson commented:
@pabigot @tejlmand Ok, pushed a rebased/updated version.

1. Defaults put in `Kconfig.zephyr`. Let me know if there is consensus on a different location, I'll happily move it.
2. Fixes up commit message based on @pabigot feedback.
3. Does *not* switch to the "friendly name" for the family ID defaults, pending some continued discussion w/ @pabigot. I don't have a strong opinion on this, just answering some follow up questions before switching. @pabigot if you still want me to switch to the name alias for those the script supports, let me know.

pabigot commented:
I'm fine with the hex constants under the argument that the canonical source for the mapping is the identified URL.  You need to add a signed-off-by to that commit.

Also my comments on commit format also applied to the third commit, which still has `feat(board)` with a list-format body.

petejohanson commented:
> I'm fine with the hex constants under the argument that the canonical source for the mapping is the identified URL. You need to add a signed-off-by to that commit.
> 
> Also my comments on commit format also applied to the third commit, which still has `feat(board)` with a list-format body.

Thanks. missed that when I was rebasing. Hopefully all good now, from a commit message perspective.

tejlmand commented:
>     1. Defaults put in `Kconfig.zephyr`. Let me know if there is consensus on a different location, I'll happily move it.

I think this is a good current solution.
I'm not really fond of any of the discussed alternative, and even though it does clutter Kconfig.zephyr, it's ok for now.

In Toolchain WG, it has been discussed to make a more proper Kconfig tree for the toolchain and compiler settings in general, and that work itself will clean up Kconfig.zephyr, so let's have the UF2 directly there, and we can see if the build outputs in general will deserve its own Kconfig.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -344,6 +344,47 @@ config BUILD_OUTPUT_S19
+if BUILD_OUTPUT_UF2
+
+config BUILD_OUTPUT_UF2_FAMILY_ID
+	# Prompt less config to setup a list of default UF2 Family IDs
```
It should no longer be prompt less, as I suppose users should still be able to provide a custom string if current soc is not known. So you probably want to re-add.

```suggestion
	string "UF2 Device Family" 
```
Note: The prompt less was only in case we wanted to have the defaults located outside of `Kconfig.zephyr`

petejohanson replied:
Ok. Prompt restored.

tejlmand reviewed (commented):
Looking good. Thanks for the patience.

Just one small comment.

tejlmand reviewed (approved):
lgtm

pabigot reviewed (approved):
Implementation looks good now; thanks.

> > AFAICT this would enable Zephyr on a large number of [Adafruit Nordic boards]
> That is correct.

For some definition of "large".   From [this](https://learn.adafruit.com/adafruit-clue/troubleshooting):
> Only Adafruit Express boards and the Trinket M0 and Gemma M0 boards ship with the UF2 bootloader installed. Feather M0 Basic, Feather M0 Adalogger, and similar boards use a regular Arduino-compatible bootloader, which does not show a boardnameBOOT drive.

So it's not clear whether this helps me.

petejohanson commented:
> Implementation looks good now; thanks.
> 
> > > AFAICT this would enable Zephyr on a large number of [Adafruit Nordic boards]
> > > That is correct.
> 
> For some definition of "large". From [this](https://learn.adafruit.com/adafruit-clue/troubleshooting):
> 
> > Only Adafruit Express boards and the Trinket M0 and Gemma M0 boards ship with the UF2 bootloader installed. Feather M0 Basic, Feather M0 Adalogger, and similar boards use a regular Arduino-compatible bootloader, which does not show a boardnameBOOT drive.
> 
> So it's not clear whether this helps me.

There's been some decent activity on https://github.com/adafruit/tinyuf2 recently, so maybe hope moving forward more of their boards will ship w/ uf2 out of the box?

MaureenHelm commented:
> > Implementation looks good now; thanks.
> > > > AFAICT this would enable Zephyr on a large number of [Adafruit Nordic boards]
> > > > That is correct.
> > 
> > 
> > For some definition of "large". From [this](https://learn.adafruit.com/adafruit-clue/troubleshooting):
> > > Only Adafruit Express boards and the Trinket M0 and Gemma M0 boards ship with the UF2 bootloader installed. Feather M0 Basic, Feather M0 Adalogger, and similar boards use a regular Arduino-compatible bootloader, which does not show a boardnameBOOT drive.
> > 
> > 
> > So it's not clear whether this helps me.
> 
> There's been some decent activity on [adafruit/tinyuf2](https://github.com/adafruit/tinyuf2) recently, so maybe hope moving forward more of their boards will ship w/ uf2 out of the box?

cc: @tannewt

pabigot commented:
> There's been some decent activity on https://github.com/adafruit/tinyuf2 recently, so maybe hope moving forward more of their boards will ship w/ uf2 out of the box?

That could help.

I really wanted Zephyr on the Clue, but it only supports DFU/serial programming, and the bootloader is placed in high flash with a configuration register overwritten to make the system start there, so the chance of bricking the device during porting is really high.  SWD pins are provided only on unlabeled test points, so there seems no solution but to throw the boards into a box or donate them to somebody who understands Arduino or CircuitPython.

Too bad: it would have been a really cool Zephyr platform.

tannewt commented:
> >   Only Adafruit Express boards and the Trinket M0 and Gemma M0 boards ship with the UF2 bootloader installed. Feather M0 Basic, Feather M0 Adalogger, and similar boards use a regular Arduino-compatible bootloader, which does not show a boardnameBOOT drive.

> So it's not clear whether this helps me.

Sorry about this text. It's old. All nRF boards from Adafruit should have a UF2 bootloader.

> There's been some decent activity on https://github.com/adafruit/tinyuf2 recently, so maybe hope moving forward more of their boards will ship w/ uf2 out of the box?

Generally we ship UF2 bootloaders on all of our boards. All SAMD51 and nRF boards will. The text above is for M0, SAMD21 boards only. Basically ones designed before UF2 may not have it. (It depends on if the test jig gets updated.)

It's less urgent for us on boards with ROM bootloaders (ESP32-S2, iMX RT, STM32) so the first batches may not have it because the UF2 bootloader isn't ready yet. Our long term goal is to always have UF2 to ensure a consistent experience for all of our boards. (Once UF2 is ready, it depends on updating the test jig.)

tinyuf2 is my hope for the future. We have separate SAMD and nRF implementations that have been a pain to keep consistent. tinyuf2 builds on top of tinyusb to bring the uf2 bootloader to all of the same platforms. (CircuitPython builds on tinyusb for the same purpose.)

> I really wanted Zephyr on the [Clue](https://www.adafruit.com/product/4500), but it only supports DFU/serial programming, and the bootloader is placed in high flash with a configuration register overwritten to make the system start there, so the chance of bricking the device during porting is really high.

Mine has a UF2 bootloader. Just double tap reset. The UF2 bootloader ([source](https://github.com/adafruit/Adafruit_nRF52_Bootloader)) may assume the soft device is present though.

If you do brick it during porting I can get you a new one.

> SWD pins are provided only on unlabeled test points, so there seems no solution but to throw the boards into a box or donate them to somebody who understands Arduino or CircuitPython.

The [Feather nRF52840](https://www.adafruit.com/product/4062) is the best board if you want 2x5 SWD. It should work similar enough to the clue if you need SWD to debug a bad crash. If you need SWD on the clue you can solder from the test points to a larger pad and then use a micro:bit breakout to get it to a bread board. Eagle files for it are here: https://github.com/adafruit/Adafruit-CLUE-PCB (To figure out which test point is which.)

You should try CircuitPython if you haven't. It's very easy to get started. This is the setup guide: learn.adafruit.com/welcome-to-circuitpython and the API overview is here: https://learn.adafruit.com/circuitpython-essentials/

> Too bad: it would have been a really cool Zephyr platform.

It would be!

tannewt commented:
> Is there an alternative to `uf2conv.py` that has any market share?

Not that I know of. We use uf2conv.py. Raspberry Pi has their own converter I think.

tannewt reviewed (approved):
I skimmed the code and it looks reasonable to me.

carlescufi commented:
Scancode error:
>Error: * scan/scripts/uf2conv.py is not apache-2.0 licensed: mit

Since this is a script that does not end in the firmware, I believe this can be considered acceptable.
